apps/web/src/components/WalletConnect/useHandleWcRequest.tsx

```diff
@@ -0,0 +1,122 @@
+
+import { BatchSignPage } from "../SendFlow/common/BatchSignPage";
+import { SingleSignPage } from "../SendFlow/common/SingleSignPage";
+import { type SdkSignPageProps, type SignHeaderProps } from "../SendFlow/utils";
+
+/**
+ * @returns a function that handles a beacon message and opens a modal with the appropriate content
+ *
+ * For operation requests it will also try to convert the operation(s) to our {@link Operation} format,
+ * estimate the fee and open the BeaconSignPage only if it succeeds
+ */
+export const useHandleWcRequest = () => {
+  const { openWith } = useDynamicModalContext();
+  const { handleAsyncActionUnsafe } = useAsyncActionHandler();
+  const getAccount = useGetOwnedAccountSafe();
+  const findNetwork = useFindNetwork();
+  const toast = useToast();
+
+  return async (
+    event: {
+      verifyContext: Verify.Context;
+    } & SignClientTypes.BaseEventArgs<{
+      request: {
+        method: string;
+        params: any;
+        expiryTimestamp?: number;
+      };
+      chainId: string;
+    }>,
+    session: SessionTypes.Struct
+  ) => {
+    await handleAsyncActionUnsafe(
+      async () => {
+        const { id, topic, params } = event;
+        const { request, chainId } = params;
+
+        let modal;
+        let onClose;
+
+        switch (request.method) {
+          case "tezos_getAccounts": {
+            const response = formatJsonRpcError(id, getSdkError("INVALID_METHOD").message);
+            await walletKit.respondSessionRequest({ topic, response });
+            return;
+          }
+
+          case "tezos_sign": {
+            // onClose = async () => {
+            //   const response = formatJsonRpcError(id, getSdkError("USER_REJECTED").message);
+            //   await walletKit.respondSessionRequest({ topic, response });
+            // };
+            // return openWith(<SignPayloadRequestModal request={"FIXME"} />, { onClose });
+            const response = formatJsonRpcError(id, getSdkError("INVALID_METHOD").message);
+            await walletKit.respondSessionRequest({ topic, response });
+            return;
+          }
+
+          case "tezos_send": {
+            if (!request.params.account) {
+              throw new Error("Missing account in request");
+            }
+            const signer = getAccount(request.params.account);
+            if (!signer) {
+              throw new Error(`Unknown account, no signer: ${request.params.account}`);
+            }
+            const operation = toAccountOperations(
+              request.params.operations,
+              signer as ImplicitAccount
+            );
+            const network = findNetwork(chainId.split(":")[1]);
+            if (!network) {
+              const response = formatJsonRpcError(id, getSdkError("INVALID_EVENT").message);
+              await walletKit.respondSessionRequest({ topic, response });
+              toast({ description: `Unsupported network: ${chainId}`, status: "error" });
```

Comment:
basically we can throw an CustomError and it should be handled inside handleAsyncActionUnsafe

Reply:
the problem is that it will send a generic reject to the peer. I need to pass my own more detailed answer so I cannot throw an error to my understanding.
Or maybe I need my own WalletConnect Error